examples/fabric-admin/device_manager/DeviceManager.cpp

```diff
@@ -0,0 +1,226 @@
+        mLastUsedNodeId = nodeId;
+    }
+}
+
+void DeviceManager::AddSyncedDevice(const Device & device)
+{
+    mSyncedDevices.insert(device);
+    ChipLogProgress(NotSpecified, "Added synced device: NodeId:" ChipLogFormatX64 ", EndpointId %u",
+                    ChipLogValueX64(device.GetNodeId()), device.GetEndpointId());
+}
+
+Device * DeviceManager::FindDeviceByEndpoint(EndpointId endpointId)
+{
+    for (auto & device : mSyncedDevices)
+    {
+        if (device.GetEndpointId() == endpointId)
+        {
+            return const_cast<Device *>(&device);
+        }
+    }
+    return nullptr;
+}
+
+Device * DeviceManager::FindDeviceByNode(NodeId nodeId)
+{
+    for (auto & device : mSyncedDevices)
+    {
+        if (device.GetNodeId() == nodeId)
+        {
+            return const_cast<Device *>(&device);
+        }
+    }
+    return nullptr;
+}
+
+void DeviceManager::RemoveSyncedDevice(NodeId nodeId)
+{
+    Device * device = FindDeviceByNode(nodeId);
+    if (device != nullptr)
+    {
+        mSyncedDevices.erase(*device);
+        ChipLogProgress(NotSpecified, "Removed synced device: NodeId:" ChipLogFormatX64 ", EndpointId %u",
+                        ChipLogValueX64(device->GetNodeId()), device->GetEndpointId());
+    }
+    else
+    {
+        ChipLogProgress(NotSpecified, "No device found with NodeId:" ChipLogFormatX64, ChipLogValueX64(nodeId));
+    }
+}
+
+void DeviceManager::HanldeAttributeChange(const app::ConcreteDataAttributePath & path, TLV::TLVReader * data)
+{
+    if (path.mClusterId != Descriptor::Id || path.mAttributeId != Descriptor::Attributes::PartsList::Id)
+    {
+        return;
+    }
+
+    ChipLogProgress(NotSpecified, "Attribute change detected:");
+    ChipLogProgress(
+        NotSpecified, "Endpoint: %u, Cluster: " ChipLogFormatMEI ", Attribute: " ChipLogFormatMEI ", DataVersion: %" PRIu32,
+        path.mEndpointId, ChipLogValueMEI(path.mClusterId), ChipLogValueMEI(path.mAttributeId), path.mDataVersion.ValueOr(0));
+
+    app::DataModel::DecodableList<EndpointId> value;
+    CHIP_ERROR error = app::DataModel::Decode(*data, value);
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(NotSpecified, "Failed to decode attribute value. Error: %" CHIP_ERROR_FORMAT, error.Format());
+        return;
+    }
+
+    std::set<EndpointId> newEndpoints;
+
+    // Populate the newEndpoints set from the decoded value using an iterator
+    auto iter = value.begin();
+    while (iter.Next())
+    {
+        newEndpoints.insert(iter.GetValue());
+    }
+
+    if (iter.GetStatus() != CHIP_NO_ERROR)
+    {
+        ChipLogError(NotSpecified, "Failed to iterate over decoded attribute value.");
+        return;
+    }
+
+    // Compare newEndpoints with mSyncedEndpoints to determine added and removed endpoints
+    std::vector<EndpointId> addedEndpoints;
+    std::vector<EndpointId> removedEndpoints;
+
+    // Find added endpoints
```

Comment:
Originally I was wondering why you are not using `std::set_intersection` and `std::set_symmetric_difference`, but I can see that the reason is because you are seaching by endpoint and not device.

I would recommend adding a comment why you use sets, but then do the search manually to not trip up future reviewer/coders that look at this in the future

Reply:
Added comment explains why we're not using set operations directly,